src/docs/05-go-client/19-workflow-non-deterministic-error.md

```diff
@@ -0,0 +1,178 @@
+```
+to
+```go
+fb := workflow.ExecuteActivity(ctx, activityB, a)
+fc := workflow.ExecuteActivity(ctx, activityC, a)
+err = fb.Get(ctx,nil)
+if err != nil {
+	return err
+}
+err = fc.Get(ctx,nil)
+if err != nil {
+	return err
+}
+```
+And restart worker, then it will run into this error. Because in the history, the workflow has scheduled only activityB after the one minute timer, however, during replay, there are two activities scheduled in a decision( in parallel).
+
+### 3. Mismatched decisions
+```go
+fmt.Errorf("nondeterministic workflow: history event is %s, replay decision is %s",util.HistoryEventToString(e), util.DecisionToString(d))
+```
+For source code click [here](https://github.com/uber-go/cadence-client/blob/e5081b085b0333bac23f198e57959681e0aee987/internal/internal_task_handlers.go#L1214)
+
+This means after replay code, the decision scheduled is different than the one in history. Using the previous history as an example, when the workflow is waiting at the one hour timer(event ID 22),
+if we change the line of :
+```go
+err = workflow.ExecuteActivity(ctx, ActivityB, a).Get(ctx, nil)
+```
+to
+```go
+err = workflow.ExecuteActivity(ctx, ActivityC, a).Get(ctx, nil)
+```
+And restart worker, then it will run into this error. Because in the history, the workflow has scheduled ActivityB with input a, but during replay, it schedules ActivityC.
+
+### 4. Decision state machine panic
+```go
+fmt.Sprintf("unknown decision %v, possible causes are nondeterministic workflow definition code"+" or incompatible change in the workflow definition", id)
+```
+For source code click [here](https://github.com/uber-go/cadence-client/blob/e5081b085b0333bac23f198e57959681e0aee987/internal/internal_decision_state_machine.go#L693)
+
+This usually means workflow history is corrupted due to some bug. For example, the same activity can be scheduled and differentiated by activityID. So ActivityIDs for different activities are supposed to be unique in workflow history. If however we have an ActivityID collision, replay will run into this error.
+
+## Common Q&A
+
+### I want to change my workflow implementation. What code changes may produce non-deterministic errors?
+
+As we discussed in previuos sections, if you changes mess up with decision tasks, then they will probably lead you to non-deterministic errors.
+These are some common changes that can be categorized by these previous 4 types mentioned above.
+
+1. Changing the order of executing Cadence defined operations, such as activities, timer, child workflows, signals, cancelRequest.
+2. Change the duration of a timer
+3. Use build-in goroutine of golang instead of using `workflow.Go`
+4. Use build-in channel of golang instead of using `workflow.Channel`
+5. Use build-in sleep function instead of using `workflow.Sleep`
+
+### What are some changes that will NOT trigger non-deterministic errors?
+
+Code changes that are free of non-deterministic erorrs normally do not involve decision tasks in Cadence.
```

Comment:
```suggestion
Code changes that are free of non-deterministic errors normally do not involve decision tasks in Cadence.
```
though tbh I'm not quite sure what this is trying to teach.  from above, _everything_ involves decision tasks, because that's what creates everything....... and you can't change decision tasks themselves at all, so....